### PR TITLE
Exercise2

### PR DESCRIPTION
**Pull request by Aimtracker**
Finished exercise 2:
- added test utils and configured linter
- implemented empty list view
- implemented button to order the list (ascending/descending)
- added tests for each component
- added and configured storybook
- added storybooks for each component

### CODE REVIEW

**Comment by steven-mi** [comment]
`webapp/sdf/src/stories` is not relevant and shouldn't be included in the project

File: .github/workflows/node.js.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [14.x]
+
```
Comment (OliKue) [inline .github/workflows/node.js.yml]: Inside the workflows its recomendet to test your code against different enviroments (or at least its very easy and it wont hurt to do so)...
```suggestion
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [windows-latest, ubuntu-20.04]
        node-version: [10.x, 12.x, 14.x]

File: webapp/sdf/.storybook/main.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    "../src/**/*.stories.@(js|jsx|ts|tsx)",
+    "../src/components/**/*.stories.js"
+  ],
+  "addons": [
+    "@storybook/addon-links",
+    "@storybook/addon-essentials",
+    '@storybook/addon-a11y',
+    '@storybook/addon-controls'
+  ]
+}
```
Comment (OliKue) [inline webapp/sdf/.storybook/main.js:13]: Just clean code improvements.
```suggestion
}

File: webapp/sdf/.storybook/preview.js [hunk 3]
```diff
@@ -0,0 +1,4 @@
+
+export const parameters = {
+  actions: { argTypesRegex: "^on[A-Z].*" },
+}
```
Comment (OliKue) [inline webapp/sdf/.storybook/preview.js:4]: Just clean code improvements.
```suggestion
}

File: webapp/sdf/src/components/item-component/item-component.spec.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        expect(component.emitted().updateEvent[0][0].votes).toEqual(1)
+    })
+
+    it('emits remove event with correct item', async () => {
+        const removeButton = component.find('.remove-button')
+        removeButton.trigger('click')
+        await component.vm.$nextTick()
+        expect(component.emitted().removeEvent[0][0]).toEqual({id: 1, title: 'Eintrag2', votes: 2})
+    })
+})
```
Comment (OliKue) [inline webapp/sdf/src/components/item-component/item-component.spec.js:61]: Just clean code improvements.
```suggestion
})

File: webapp/sdf/src/components/item-component/item-component.stories.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    props: Object.keys(argTypes),
+    components: { ItemComponent },
+    template: '<ItemComponent :item="$props.item" @removeEvent="removeItem" @updateEvent="updateItem"/>',
+    
+});
+  
+  export const Base = Template.bind({});
+  Base.args = {
+      item: new Item(1, 'VueJS', 3)
+  };
```
Comment (OliKue) [inline webapp/sdf/src/components/item-component/item-component.stories.js:23]: Just clean code improvements.
```suggestion
  };

File: webapp/sdf/src/components/item-form-component/item-form-component.spec.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    })
+
+    it('submit event not emitted when title == ""', async () => {
+        await component.setData({title:""})
+        const submitButton = component.find('.submit-button')
+        submitButton.trigger('submit')
+        await component.vm.$nextTick()
+        expect(component.emitted().submitEvent).not.toBeTruthy()
+    })
+})
```
Comment (OliKue) [inline webapp/sdf/src/components/item-form-component/item-form-component.spec.js:36]: Just clean code improvements.
```suggestion
})

File: webapp/sdf/src/components/item-form-component/item-form-component.stories.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  };
+
+  const Template = (args, { argTypes }) => ({
+    props: Object.keys(argTypes),
+    components: { ItemFormComponent },
+    template: '<ItemFormComponent @submitEvent="onSubmit"/>',
+    
+});
+  
+  export const Base = Template.bind({});
```
Comment (OliKue) [inline webapp/sdf/src/components/item-form-component/item-form-component.stories.js:18]: Just clean code improvements.
```suggestion
  export const Base = Template.bind({});

File: webapp/sdf/src/components/item-form-component/item-form-component.vue [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  methods: {
+    onSubmit() {
+      if (this.title != ""){
+        this.$emit("submitEvent", this.title);
+        this.title = "";
+      }
+    },
+  },
+};
+</script>
```
Comment (OliKue) [inline webapp/sdf/src/components/item-form-component/item-form-component.vue:27]: Just clean code improvements.
```suggestion
</script>

File: webapp/sdf/src/components/list-component/list-component.spec.js [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+                    await component.vm.$nextTick()
+                    toggleButton.trigger('click')
+                    await component.vm.$nextTick()
+                    const firstElement = component.findComponent(ItemComponent)
+                    expect(firstElement.text()).toContain('Eintrag1')
+                })
+            })
+        })
+    })
+})
```
Comment (OliKue) [inline webapp/sdf/src/components/list-component/list-component.spec.js:141]: Just clean code improvements.
```suggestion
})

File: webapp/sdf/src/components/list-component/list-component.stories.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    component: ListComponent,
+  };
+
+  const Template = () => ({
+    components: { ListComponent },
+    template: '<ListComponent />',
+    
+});
+  
+  export const Base = Template.bind({});
```
Comment (OliKue) [inline webapp/sdf/src/components/list-component/list-component.stories.js:14]: Just clean code improvements.
```suggestion
  export const Base = Template.bind({});

**Review by OliKue — APPROVED** [review]
Just some minor changes for clean code and workflow usage.

**Review by steven-mi — APPROVED** [review]
Suggested changes:
- `webapp/sdf/src/stories` is not relevant and shouldn't be included in the project

**Comment by MatzeK105** [comment]
We reviewed [Pull Request of Abteilung6](https://github.com/Systems-Development-and-Frameworks/abteilung6/pull/4#pullrequestreview-527511952).

File: .github/workflows/node.js.yml [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+jobs:
+  build:
+
+    runs-on: ${{ matrix.os }}
+
+    strategy:
+      matrix:
+        os: [windows-latest, ubuntu-20.04]
+        node-version: [10.x, 12.x, 14.x]
```
Comment (roschaefer) [inline .github/workflows/node.js.yml:23]: You are transpiling JavaScript, if you want your build server to ship your code, then only one operating system and one node version is needed.

File: webapp/sdf/.gitignore [hunk 12]
```diff
@@ -21,3 +21,5 @@ pnpm-debug.log*
 *.njsproj
 *.sln
 *.sw?
+
+/src/stories/
```
Comment (roschaefer) [inline webapp/sdf/.gitignore:25]:
```suggestion
```
Just remove the example code. You can also `git add -p` to selectively stage only the desired changes.

File: webapp/sdf/.storybook/main.js [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+  "stories": [
+    "../src/**/*.stories.mdx",
+    "../src/**/*.stories.@(js|jsx|ts|tsx)",
+    "../src/components/**/*.stories.js"
+  ],
+  "addons": [
+    "@storybook/addon-links",
+    "@storybook/addon-essentials",
+    '@storybook/addon-a11y',
+    '@storybook/addon-controls'
```
Comment (roschaefer) [inline webapp/sdf/.storybook/main.js:11]:
```suggestion
```
Is part of `addon-essentials`.

File: webapp/sdf/src/components/item-component/item-component.spec.js [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+    beforeEach(() => {
+        component = mount(ItemComponent, {
+            propsData: {
+                item : new Item(1, "Eintrag2", 2)
+            }
+        });
+    })
+
+    it('contains item title and votes Expected --> Eintrag (2)', () => {
```
Comment (roschaefer) [inline webapp/sdf/src/components/item-component/item-component.spec.js:16]:
```suggestion
    it('displays item title and votes, e.g `Eintrag (2)`', () => {
```

File: webapp/sdf/src/components/item-component/item-component.spec.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+        expect(component.text()).toContain('Eintrag2 (2)')
+    })
+
+    it('emits remove event', async () => {
+        const removeButton = component.find('.remove-button')
+        removeButton.trigger('click')
+        await component.vm.$nextTick()
```
Comment (roschaefer) [inline webapp/sdf/src/components/item-component/item-component.spec.js:23]:
```suggestion
        await component.find('.remove-button').trigger('click')
```

File: webapp/sdf/src/components/item-component/item-component.spec.js [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        expect(component.emitted().removeEvent).toBeTruthy()
```
Comment (roschaefer) [inline webapp/sdf/src/components/item-component/item-component.spec.js:24]:
```suggestion
        expect(component.emitted('removeEvent')).toBeTruthy()
```
https://vue-test-utils.vuejs.org/api/wrapper/emitted.html#emitted

File: webapp/sdf/src/components/item-component/item-component.spec.js [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+        await component.vm.$nextTick()
+        expect(component.emitted().updateEvent).toBeTruthy()
+    })
+
+    it('emits update event -1', async () => {
+        const downvoteButton = component.find('.downvote-button')
+        downvoteButton.trigger('click')
+        await component.vm.$nextTick()
+        expect(component.emitted().updateEvent).toBeTruthy()
+    })
```
Comment (roschaefer) [inline webapp/sdf/src/components/item-component/item-component.spec.js:39]: Test case description and test case implementation do not match. You would need to assert this:
```suggestion
    it('emits update event +1', async () => {
        const upvoteButton = component.find('.upvote-button')
        upvoteButton.trigger('click')
        await component.vm.$nextTick()
        expect(component.emitted('updateEvent').toMatchObject({votes: 3})
    })

    it('emits update event -1', async () => {
        const downvoteButton = component.find('.downvote-button')
        downvoteButton.trigger('click')
        await component.vm.$nextTick()
        expect(component.emitted('updateEvent').toMatchObject({votes: 1})
    })
```

File: webapp/sdf/src/components/item-component/item-component.spec.js [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+        expect(component.emitted().updateEvent[0][0].votes).toEqual(3)
+    })
+
+    it('updates votes -1', async () => {
+        const downvoteButton = component.find('.downvote-button')
+        downvoteButton.trigger('click')
+        await component.vm.$nextTick()
+        expect(component.emitted().updateEvent[0][0].votes).toEqual(1)
+    })
+
```
Comment (roschaefer) [inline webapp/sdf/src/components/item-component/item-component.spec.js:54]: Ah, ok, then I would merge these test cases.

File: webapp/sdf/src/components/item-component/item-component.spec.js [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    it('emits remove event with correct item', async () => {
+        const removeButton = component.find('.remove-button')
+        removeButton.trigger('click')
+        await component.vm.$nextTick()
+        expect(component.emitted().removeEvent[0][0]).toEqual({id: 1, title: 'Eintrag2', votes: 2})
```
Comment (roschaefer) [inline webapp/sdf/src/components/item-component/item-component.spec.js:59]:
```suggestion
        expect(component.emitted('removeEvent')).toEqual([[{id: 1, title: 'Eintrag2', votes: 2}]])
```
If you're not emitting so many events in the test case, I find it safer to expect the entire array.

File: webapp/sdf/src/components/item-form-component/item-form-component.spec.js [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+import ItemFormComponent from 'components/item-form-component/item-form-component.vue'
+
+describe('item-form-component', () => {
+    let component
+
+    beforeEach(() => {
+        component = mount(ItemFormComponent, {
+            data() {return {
+                title: "test"
+            }}
```
Comment (roschaefer) [inline webapp/sdf/src/components/item-form-component/item-form-component.spec.js:11]: You shouldn't need to set internal `data()` in your tests.

File: webapp/sdf/src/components/list-component/list-component.spec.js [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+                            new Item(3, "Eintrag3", -2)
+                        ]
+                    }}
+                })
+            })
+
+            describe('renders ItemComponent for each item', () => {
+                it('renders 3 components', async () => {
+                    const itemComponents = component.findAllComponents(ItemComponent)
+                    expect(itemComponents.length).toBe(3)
```
Comment (roschaefer) [inline webapp/sdf/src/components/list-component/list-component.spec.js:36]:
```suggestion
                    expect(itemComponents).toHaveLength(3)
```

File: webapp/sdf/src/components/list-component/list-component.spec.js [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+                })
+            })
+
+            describe('receives remove event for item 1', () => {
+                it('removes item 1', async () => {
+                    component.findComponent(ItemComponent).vm.$emit('removeEvent', {
```
Comment (roschaefer) [inline webapp/sdf/src/components/list-component/list-component.spec.js:42]: Mocking the event is required when you use `shallowMount` version. If you're using `mount`, I would interact with sub-components in a non-mocked way (find a particular button, click button etc.).

File: webapp/sdf/src/components/list-component/list-component.spec.js [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+                        id:1, title:"Eintrag1", votes:5
+                    })
+                    await component.vm.$nextTick()
+                    expect(component.vm.$data.itemList.some(item =>
+                        item.id == 1 &&
+                        item.title =="Eintrag1" &&
+                        item.votes == 5
+                    )).toBe(false)
```
Comment (roschaefer) [inline webapp/sdf/src/components/list-component/list-component.spec.js:50]: This is too error-prone. I would suggest `Arrange - Assert - Act- Assert`:
```js
expect(component.findAllComponents).toHaveLength(3)
await component.find('button.remove').trigger('click')
expect(component.findAllComponents).toHaveLength(2)
```

Comment (roschaefer) [reply webapp/sdf/src/components/list-component/list-component.spec.js:50]: By the way, I just found: https://stackoverflow.com/questions/1021007/should-it-be-arrange-assert-act-assert

I like the way of calling it: Arrange - *Assume* - Act - Assert

File: webapp/sdf/src/components/list-component/list-component.spec.js [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+                    })
+                    await component.vm.$nextTick()
+                    expect(component.vm.$data.itemList.some(item =>
+                        item.id == 2 &&
+                        item.title =="Eintrag2" &&
+                        item.votes == 2
+                    )).toBe(true)
+                })
+            
+                it('does not remove item 3', async () => {
```
Comment (roschaefer) [inline webapp/sdf/src/components/list-component/list-component.spec.js:65]: I wouldn't test this specifically. If you want, then do it in one test case:
```js
expect(component.findAllComponent(ItemComponent).wrapper.map(w => w.props())).toEqual([
  expect.objectContaining({ id: 1 })
  expect.objectContaining({ id: 2 })
  expect.objectContaining({ id: 3 })
])
await wrapper.findAll('button.remove').get(1).trigger('click')
expect(component.findAllComponent(ItemComponent).wrapper.map(w => w.props())).toEqual([
  expect.objectContaining({ id: 1 })
  expect.objectContaining({ id: 3 })
])
```

File: webapp/sdf/src/components/list-component/list-component.spec.js [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+                })
+
+                it('was descending, now ascending', async () => {
+                    const toggleButton = component.find('.toggle-btn')
+                    toggleButton.trigger('click')
+                    await component.vm.$nextTick()
+                    toggleButton.trigger('click')
+                    await component.vm.$nextTick()
+                    const firstElement = component.findComponent(ItemComponent)
+                    expect(firstElement.text()).toContain('Eintrag1')
```
Comment (roschaefer) [inline webapp/sdf/src/components/list-component/list-component.spec.js:136]: You can assert the entire list:
```js
expect(component.findAllComponent(ItemComponent).wrapper.map(w => w.props('votes'))).toEqual([3,2,1])
await component.find('.toggle-btn').trigger('click')
expect(component.findAllComponent(ItemComponent).wrapper.map(w => w.props('votes'))).toEqual([1,2,3])

```

File: webapp/sdf/src/components/list-component/list-component.spec.js [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+                it('does not remove item 3', async () => {
+                    component.findComponent(ItemComponent).vm.$emit('updateEvent', new Item(1, "Eintrag1", 6))
+                    await component.vm.$nextTick()
+                    expect(component.vm.$data.itemList.some(item =>
+                        item.id == 3 &&
+                        item.title =="Eintrag3" &&
+                        item.votes == -2
+                    )).toBe(true)
+                })
+            })
```
Comment (roschaefer) [inline webapp/sdf/src/components/list-component/list-component.spec.js:118]: I would suggest to write one test case for all of this and not assert `component.vm.$data`. Instead, assert something you can "see" (ie. the rendered `html()`).

File: webapp/sdf/src/components/list-component/list-component.vue [hunk 27]
```diff
@@ -59,7 +68,12 @@ export default {
   computed: {
     sortedArray() {
       var newArray = [...this.itemList];
-      return newArray.sort((a, b) => b.votes - a.votes);
+      return newArray.sort((a, b) => {
+        if (this.orderAscending)
+          return b.votes - a.votes
+        else
+          return a.votes - b.votes
```
Comment (roschaefer) [inline webapp/sdf/src/components/list-component/list-component.vue:75]:
```suggestion
        if (this.orderAscending) return b.votes - a.votes
        return a.votes - b.votes
```

That's a guard-clause :point_up: which is always easier to read than adding curly braces and nesting.

File: webapp/sdf/yarn.lock [hunk 28]
```diff
@@ -0,0 +1,12714 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (roschaefer) [inline webapp/sdf/yarn.lock:1]: Check *either* `yarn.lock` *or* `package-lock.json` into version control.

**Review by roschaefer — APPROVED** [review]
excellent

Excellent! Here's what I've found:

:star: Setup linter
:star: Setup jest
:star: Feature: Toggle ascending/descending order
:star: Feature: Empty state
:star: Co-locate stories, tests and component code
:star: Setup build server
:star: Setup storybook
:star: Storybook interactively controllable
:star: Storybook passes a11y tests
:star: Give a code review

You achieved 10/10 :star: congratulations!

See my suggestions below :point_down: mostly test case granularity, what to test
and what not to test. Apart from that, this PR is quite perfect, good job!

I especially like the readability of your tests:
```js
  list-component
    item list
      empty
        ✓ shows message "The list is empty!" (27 ms)
      not empty
        renders ItemComponent for each item
          ✓ renders 3 components (16 ms)
```

**Pull request merged**